src/SourceLocator/Ast/FindReflectionsInTree.php

```diff
@@ -41,94 +43,64 @@ public function __invoke(
         IdentifierType $identifierType,
         LocatedSource $locatedSource
     ) : array {
-        $nodeVisitor = new class($reflector, $identifierType, $locatedSource, $this->astConversionStrategy)
-            extends NodeVisitorAbstract
-        {
-            /** @var Reflection[] */
-            private $reflections = [];
-
-            /** @var Reflector */
-            private $reflector;
-
-            /** @var IdentifierType */
-            private $identifierType;
-
-            /** @var LocatedSource */
-            private $locatedSource;
-
-            /** @var AstConversionStrategy */
-            private $astConversionStrategy;
-
-            /** @var Namespace_|null */
-            private $currentNamespace;
-
-            public function __construct(
-                Reflector $reflector,
-                IdentifierType $identifierType,
-                LocatedSource $locatedSource,
-                AstConversionStrategy $astConversionStrategy
-            ) {
-                $this->reflector             = $reflector;
-                $this->identifierType        = $identifierType;
-                $this->locatedSource         = $locatedSource;
-                $this->astConversionStrategy = $astConversionStrategy;
-            }
-
-            public function enterNode(Node $node) : void
-            {
-                if ($node instanceof Namespace_) {
-                    $this->currentNamespace = $node;
-
-                    return;
-                }
 
-                if ($node instanceof Node\Stmt\ClassLike) {
-                    $classNamespace = $node->name === null ? null : $this->currentNamespace;
-                    $reflection     = $this->astConversionStrategy->__invoke($this->reflector, $node, $this->locatedSource, $classNamespace);
-
-                    if ($this->identifierType->isMatchingReflector($reflection)) {
-                        $this->reflections[] = $reflection;
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor(new NameResolver());
+        $ast = $nodeTraverser->traverse($ast);
+
+        static $collectNamespaceOrphanSymbols;
+        static $collectNamespacedSymbols;
+
+        if($collectNamespacedSymbols === null){
+
+            $collectNamespaceOrphanSymbols = [
+                // do not enter Namespace_ but do collect it
+                new CollectByTypeInstructions(Namespace_::CLASS, true, []),
+
+                // collect ClassLike and keep searching in its stmts
+                new CollectByTypeInstructions(ClassLike::CLASS, true, ['stmts']),
+
+                // collect Function_ and keep searching in its stmts
+                new CollectByTypeInstructions(Function_::CLASS, true, ['stmts']),
+
+                // @TODO can add more instructions to blacklist bits of the AST where not to
```

Comment:
If I understand this correctly, we look for anything that isn't a block (anything that supports `{}`), a function, a method or a class, but we don't include the non-class and non-function ones by using `false` as second constructor parameter of `CollectByTypeInstructions`?

Reply:
Pretty much overrides $node->getSubNodesNames(). so for example in classes rather than looking in all sub nodes, it just just looks in "stmts". For unspecified types, it fallbacks to $node->getSubNodesNames().

Reply:
Didn't yet try it out, but it does look indeed much lighter. Yes, the "tail recursion via GOTO" helps :D